instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/snippet/InjectionState.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.servlet.v3_0.snippet;
+
+// this is shared by both ServletOutputStream and PrintWriter injection
+public class InjectionState {
+  private static final int HEAD_TAG_WRITTEN_FAKE_VALUE = -1;
+  private static final int HEAD_TAG_LENGTH = "<head>".length();
+  private final SnippetInjectingResponseWrapper wrapper;
+  private int headTagBytesSeen = 0;
+
+  public InjectionState(SnippetInjectingResponseWrapper wrapper) {
+    this.wrapper = wrapper;
+  }
+
+  public int getHeadTagBytesSeen() {
+    return headTagBytesSeen;
+  }
+
+  public String getCharacterEncoding() {
+    return wrapper.getCharacterEncoding();
+  }
+
+  public void setHeadTagWritten() {
+    headTagBytesSeen = HEAD_TAG_WRITTEN_FAKE_VALUE;
+  }
+
+  public boolean isHeadTagWritten() {
+    return headTagBytesSeen == HEAD_TAG_WRITTEN_FAKE_VALUE;
+  }
+
+  /**
+   * Returns true when the byte is the last character of "<head>" and now is the right time to
+   * inject. Otherwise, returns false.
+   */
+  public boolean processByte(int b) {
+    if (isHeadTagWritten()) {
+      return false;
+    }
+    if (inHeadTag(b)) {
+      headTagBytesSeen++;
+    } else {
+      headTagBytesSeen = 0;
+    }
+    return headTagBytesSeen == HEAD_TAG_LENGTH;
```

Comment:
You don't need to expose the `setHeadTagWritten`: you can encapsulate it inside this method and make this class more failure-proof